File: java/lib/src/main/java/com/zourzouvillys/httpsig/Algorithms.java
package com.zourzouvillys.httpsig;

import java.security.GeneralSecurityException;
import java.security.MessageDigest;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.MGF1ParameterSpec;
import java.security.spec.PSSParameterSpec;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * Low-level crypto operations for each RFC 9421 algorithm.
 *
 * Not public API, just the plumbing that {@link Keys} delegates to.
 */
final class Algorithms {

    private Algorithms() {}

    // ---- RSA-PSS-SHA512 ----

    private static final PSSParameterSpec PSS_PARAMS = new PSSParameterSpec(
        "SHA-512", "MGF1", MGF1ParameterSpec.SHA512, 64, 1
    );

    static byte[] rsaPssSign(PrivateKey key, byte[] data) throws HttpSigException {
        try {
            Signature sig = Signature.getInstance("RSASSA-PSS");
            sig.setParameter(PSS_PARAMS);
            sig.initSign(key);
            sig.update(data);
            return sig.sign();
        } catch (GeneralSecurityException e) {
            throw new HttpSigException("RSA-PSS sign failed", e);
        }
    }

    static boolean rsaPssVerify(PublicKey key, byte[] data, byte[] signature) throws HttpSigException {
        try {
            Signature sig = Signature.getInstance("RSASSA-PSS");
            sig.setParameter(PSS_PARAMS);
            sig.initVerify(key);
            sig.update(data);
            return sig.verify(signature);
        } catch (GeneralSecurityException e) {
            throw new HttpSigException("RSA-PSS verify failed", e);
        }
    }

    // ---- ECDSA P-256 SHA-256 ----
    // RFC 9421 requires raw r||s (64 bytes). Java 15+ has SHA256withECDSAinP1363Format
    // which produces this format directly.

    private static final String ECDSA_P1363 = "SHA256withECDSAinP1363Format";

    static byte[] ecdsaSign(PrivateKey key, byte[] data) throws HttpSigException {
        try {
            Signature sig = Signature.getInstance(ECDSA_P1363);
            sig.initSign(key);
            sig.update(data);
            return sig.sign();
        } catch (GeneralSecurityException e) {
            throw new HttpSigException("ECDSA sign failed", e);
        }
    }

    static boolean ecdsaVerify(PublicKey key, byte[] data, byte[] signature) throws HttpSigException {
        try {
            // the signature should be exactly 64 bytes (r||s, 32 each)
            if (signature.length != 64) {
                return false;
            }
            Signature sig = Signature.getInstance(ECDSA_P1363);
            sig.initVerify(key);
            sig.update(data);
            return sig.verify(signature);
        } catch (GeneralSecurityException e) {
            throw new HttpSigException("ECDSA verify failed", e);
        }
    }

    // ---- Ed25519 ----

    static byte[] ed25519Sign(PrivateKey key, byte[] data) throws HttpSigException {
        try {
            Signature sig = Signature.getInstance("Ed25519");
            sig.initSign(key);
            sig.update(data);
            return sig.sign();
        } catch (GeneralSecurityException e) {
            throw new HttpSigException("Ed25519 sign failed", e);
        }
    }

    static boolean ed25519Verify(PublicKey key, byte[] data, byte[] signature) throws HttpSigException {
        try {
            Signature sig = Signature.getInstance("Ed25519");
            sig.initVerify(key);
            sig.update(data);
            return sig.verify(signature);
        } catch (GeneralSecurityException e) {
            throw new HttpSigException("Ed25519 verify failed", e);
        }
    }

    // ---- HMAC-SHA256 ----

    static byte[] hmacSign(byte[] secret, byte[] data) throws HttpSigException {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secret, "HmacSHA256"));
            return mac.doFinal(data);
        } catch (GeneralSecurityException e) {
            throw new HttpSigException("HMAC-SHA256 sign failed", e);
        }
    }

    static boolean hmacVerify(byte[] secret, byte[] data, byte[] signature) throws HttpSigException {
        byte[] expected = hmacSign(secret, data);
        return MessageDigest.isEqual(expected, signature);
    }
}

File: kotlin/lib/src/main/kotlin/com/zourzouvillys/httpsig/Algorithms.kt
package com.zourzouvillys.httpsig

import java.security.GeneralSecurityException
import java.security.MessageDigest
import java.security.PrivateKey
import java.security.PublicKey
import java.security.Signature
import java.security.spec.MGF1ParameterSpec
import java.security.spec.PSSParameterSpec
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Low-level crypto operations for each RFC 9421 algorithm.
 *
 * Not public API, just the plumbing that [Keys] delegates to.
 */
internal object Algorithms {

    // ---- RSA-PSS-SHA512 ----

    private val PSS_PARAMS = PSSParameterSpec(
        "SHA-512", "MGF1", MGF1ParameterSpec.SHA512, 64, 1
    )

    fun rsaPssSign(key: PrivateKey, data: ByteArray): ByteArray = try {
        Signature.getInstance("RSASSA-PSS").apply {
            setParameter(PSS_PARAMS)
            initSign(key)
            update(data)
        }.sign()
    } catch (e: GeneralSecurityException) {
        throw HttpSigException("RSA-PSS sign failed", e)
    }

    fun rsaPssVerify(key: PublicKey, data: ByteArray, signature: ByteArray): Boolean = try {
        Signature.getInstance("RSASSA-PSS").apply {
            setParameter(PSS_PARAMS)
            initVerify(key)
            update(data)
        }.verify(signature)
    } catch (e: GeneralSecurityException) {
        throw HttpSigException("RSA-PSS verify failed", e)
    }

    // ---- ECDSA P-256 SHA-256 ----
    // RFC 9421 requires raw r||s (64 bytes). Java 15+ has SHA256withECDSAinP1363Format
    // which produces this format directly.

    private const val ECDSA_P1363 = "SHA256withECDSAinP1363Format"

    fun ecdsaSign(key: PrivateKey, data: ByteArray): ByteArray = try {
        Signature.getInstance(ECDSA_P1363).apply {
            initSign(key)
            update(data)
        }.sign()
    } catch (e: GeneralSecurityException) {
        throw HttpSigException("ECDSA sign failed", e)
    }

    fun ecdsaVerify(key: PublicKey, data: ByteArray, signature: ByteArray): Boolean {
        if (signature.size != 64) return false
        return try {
            Signature.getInstance(ECDSA_P1363).apply {
                initVerify(key)
                update(data)
            }.verify(signature)
        } catch (e: GeneralSecurityException) {
            throw HttpSigException("ECDSA verify failed", e)
        }
    }

    // ---- Ed25519 ----

    fun ed25519Sign(key: PrivateKey, data: ByteArray): ByteArray = try {
        Signature.getInstance("Ed25519").apply {
            initSign(key)
            update(data)
        }.sign()
    } catch (e: GeneralSecurityException) {
        throw HttpSigException("Ed25519 sign failed", e)
    }

    fun ed25519Verify(key: PublicKey, data: ByteArray, signature: ByteArray): Boolean = try {
        Signature.getInstance("Ed25519").apply {
            initVerify(key)
            update(data)
        }.verify(signature)
    } catch (e: GeneralSecurityException) {
        throw HttpSigException("Ed25519 verify failed", e)
    }

    // ---- HMAC-SHA256 ----

    fun hmacSign(secret: ByteArray, data: ByteArray): ByteArray = try {
        Mac.getInstance("HmacSHA256").apply {
            init(SecretKeySpec(secret, "HmacSHA256"))
        }.doFinal(data)
    } catch (e: GeneralSecurityException) {
        throw HttpSigException("HMAC-SHA256 sign failed", e)
    }

    fun hmacVerify(secret: ByteArray, data: ByteArray, signature: ByteArray): Boolean {
        val expected = hmacSign(secret, data)
        return MessageDigest.isEqual(expected, signature)
    }
}
